### PR TITLE
PointsProcessor can now ask for another pass over the data.

### PR DESCRIPTION
**Pull request by SirVer**
Also added a function that can remove points.

### CODE REVIEW

File: cartographer/io/points_processor.h [hunk 1]
```diff
@@ -23,18 +23,26 @@ namespace cartographer {
...
+
   PointsProcessor() {}
   virtual ~PointsProcessor() {}
 
   PointsProcessor(const PointsProcessor&) = delete;
   PointsProcessor& operator=(const PointsProcessor&) = delete;
 
-  virtual void Process(const PointsBatch& points_batch) = 0;
-  virtual void Flush() = 0;
+  // Receive a batch of 'points', process it and pass it on.
```
Comment (damonkohler) [inline cartographer/io/points_processor.h]: Fix comment.

File: cartographer/io/xray_points_processor.cc [hunk 2]
```diff
@@ -83,17 +83,24 @@ XRayPointsProcessor::XRayPointsProcessor(const double voxel_size,
       transform_(transform),
       voxels_(voxel_size, Eigen::Vector3f::Zero()) {}
 
-void XRayPointsProcessor::Process(const PointsBatch& batch) {
+void XRayPointsProcessor::Process(PointsBatch batch) {
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: const ref?

**Comment by SirVer** [comment]
@damonkohler ptal, this grew a bit.

File: cartographer/io/xray_points_processor.cc [hunk 3]
```diff
@@ -83,17 +83,24 @@ XRayPointsProcessor::XRayPointsProcessor(const double voxel_size,
...
+  next_->Process(std::move(batch));
 }
 
-void XRayPointsProcessor::Flush() {
+PointsProcessor::FlushResult XRayPointsProcessor::Flush() {
   WriteImage();
-  return next_->Flush();
+  switch (next_->Flush()) {
+    case FlushResult::kRestartStream:
+      LOG(FATAL) << "Not supported";
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: "X-Ray generation must be configured to occur after any stages that require multiple passes."

File: cartographer/io/xray_points_processor.cc [hunk 4]
```diff
@@ -83,17 +83,25 @@ XRayPointsProcessor::XRayPointsProcessor(const double voxel_size,
...
 }
 
-void XRayPointsProcessor::Flush() {
+PointsProcessor::FlushResult XRayPointsProcessor::Flush() {
   WriteImage();
-  return next_->Flush();
+  switch (next_->Flush()) {
+    case FlushResult::kRestartStream:
+      LOG(FATAL) << "X-Ray generation must be configured to occur after any "
+                    "stages that require multiple passes";
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: Missing punctuation.

Comment (SirVer) [inline cartographer/io/xray_points_processor.cc]: will fix in next PR.

**Pull request merged**